docs/design/storagenode-graceful-exit-protocol.md

```diff
@@ -0,0 +1,215 @@
+# Storage Node Graceful Exit - Transferring Pieces
+
+## Abstract
+
+This document describes how storage node transfers its pieces during Graceful Exit.
+
+## Background
+
+During Graceful Exit a storage node needs to transfer pieces to other nodes. During transfering the storage node or satellite may crash, hence it needs to be able to continue after a restart. 
+
+Satellite gathers transferred pieces list asynchronously, which is described in [Gathering Pieces Document](storagenode-graceful-exit-pieces.md). This may consume a significant amount of time.
+
+Transferring a piece to another node may fail, hence we need to ensure that critical pieces get transferred. Storage nodes can be malicious and try to misreport transfer as "failed" or "completed". A storage node may also try to send incorrect data. This means we need proof that the correct piece was transferred.
+
+After all pieces have been transferred the storage node needs a receipt for completing the transfer.
+
+Both storage node and satellite operators need insight into the Graceful Exit progress.
+
+## Design
+
+The storage node has a Graceful Exit service, which ensures that the process is completed. It queries `graceful_exit_status` to find unfinished exits and starts a per satellite `worker` if needed.
+
+The `worker` polls the satellite, requesting pieces to transfer. The satellite will initiate a Graceful Exit if not already initiated. When first initiated, the satellite will start gathering pieces for the exiting node and return `NotReady`. The satellite will continue to return `NotReady` until the piece gathering process has completed.
+
+The `worker` should continue to poll the satellite at a configurable interval until it returns pieces to transfer.  
+
+The storage node should concurrently transfer pieces returned by the satellite. The storage node should send a `TransferSucceeded` message as pieces are successfuly transfered. The Storage node should send a `TransferFailed`, with reason, on failure.
+
+The satellites should set the `finished_at` on success, and respond with a `DeletePiece` message. Otherwise set the `failed_at` and `failure_status_code` for reprocessing.
+
+The satellite should respond with a `ExitCompleted` message when all pieces have finished processing. 
+The storage node should store the completion receipt, stop transfer processing, and remove all remaining pieces for the satellite.
+
+When the storage node has failed too many transfers or has sent incorrect data, the satellite will send `ExitFailed`. This indicates that the process has ended ungracefully.
```

Comment:
I think the satellite should send a completion receipt in this case too, with failed status and reason for the failure.

Reply:
Updated